### DISCUSSION TITLE
Would it be better to disable the "blunderbuss" plugin?

### DISCUSSION THREAD

**Discussion by bobcatfish** [tektoncd/community#362] [Ideas]
We have the prow `blunderbuss` plugin enabled for all projects (https://github.com/tektoncd/plumbing/blob/894797a2511178fd0e8f43e82e131983144bc28e/prow/plugins.yaml#L25) but for the projects that I've worked on it seems like we consistently ignore it, and in fact for TEPs in this repo, we're using assignees where we probably want to use reviewers because the reviewers list is noisy and not useful.

I'm wondering if we might be better off without it (at least for now)? We can always keep it on for any projects that are using it.

@afrittoli @vdemeester what do you think?

**Comment by bobcatfish** [comment]
(btw details about our enabled plugins and what they do are available at https://prow.tekton.dev/plugins)

**Reply by vdemeester** [reply]
On a side not, there is a bunch of plugins that seem to be enabled that we could disable : `bugzilla`, `buildifier`, `golint`, `stage` and *maybe* `joke` and `project` (as we don't use project in any *project* so far)

**Comment by vdemeester** [comment]
So `blunderbuss` is the *automation* of the  `assign` plugin. `assign` allows us to us `cc` and `assign` command, and `blunderbuss` does "pick" some up from the OWNERs.

> The blunderbuss plugin automatically requests reviews from reviewers when a new PR is created. The reviewers are selected based on the reviewers specified in the OWNERS files that apply to the files modified by the PR.

This allows us to even "pick" the reviewers from a specific set of people, if we use folder OWNER (as we do [here](https://github.com/tektoncd/pipeline/blob/master/pkg/apis/OWNERS) in `tektoncd/pipeline`)

![2021-03-01-120655](https://user-images.githubusercontent.com/6508/109489081-48c0e080-7a7e-11eb-9eed-af5bfdaee720.png)

This is what happens on a default PR on `tektoncd/pipeline` so I wouldn't say that the project consistently ignore it. It sets a default set of reviewers to the PRs that can then review or ping/assign other contributors/maintainers. I found it very useful to be honest 😛

**Reply by bobcatfish** [reply]
@vdemeester do you find that folks are actually paying attention to what they are assigned to review? i know i'm not 😬 

![image](https://user-images.githubusercontent.com/432502/109856805-e73c7580-7c27-11eb-986b-3b58b0c00f0f.png)

If I'm the only one ignoring it then I'm happy to change what I'm doing - but I'm under the impression this is mostly going unused

For example [this PR in pipelines](https://github.com/tektoncd/pipeline/pull/3802):

![image](https://user-images.githubusercontent.com/432502/109856980-22d73f80-7c28-11eb-89d0-3906d70a324d.png)

Or [this one](https://github.com/tektoncd/pipeline/pull/3751):

![image](https://user-images.githubusercontent.com/432502/109857036-34b8e280-7c28-11eb-90bf-1dbe92b3f757.png)

Or in community https://github.com/tektoncd/community/pull/359:

![image](https://user-images.githubusercontent.com/432502/109857092-49957600-7c28-11eb-8b50-dfab2f349614.png)

(btw im not saying we disable the "assign" plugin, just "blunderbuss")

Or maybe the answer is to clean up the owners lists and start asking all owners to pay attention to what they are assigned?

**Reply by vdemeester** [reply]
So I am *definitely* not ignoring them at least (because they appear in my notification as request review, I treat them before the others).

**Comment by vdemeester** [comment]
A bit of update : 
- `/auto-cc` comes from `blunderbuss` but does it “manually”/on-demand (request reviewers)
- `/uncc @pseudo` to remove yourself from the requested reviewer

`/cc` and `/un-cc` are coming from the `assign` plugins, so if we disable the blunderbuss plugin, we can still do this “on-demand”. One positive aspect of `blunderbuss` is that, at least some collaborators should get the *ping* and re-direct to the appropriate person if need be. That *potentially* reduces the triage work. That said, if most of the requested reviews do not do that work, `blunderbuss` is *moo* and could be uninstalled yes.

**Reply by vdemeester** [reply]
Ah wait, `/auto-cc` comes from `blunderbuss`…

**Comment by wlynch** [comment]
I like auto-assigned PR assignees/reviewers - it helps maintainers distribute the review load, and gives contributors a point of contact for their PR. My worry if we remove the blunderbuss plugin is that PRs won't be looked at and/or we put a manual burden on a small group of people to triage and assign PRs as they come in.

As a reviewer I tend to struggle with 2 things with how auto-assigned reviewers work today:
1. What is the responsibility when I get assigned as a reviewer for something I don't have a lot of knowledge with. I think something that would be useful is a counterpart to https://github.com/tektoncd/community/blob/main/process.md#assigning-yourself-an-issue to detail how to remove yourself. This might also be a signal that auto assignments aren't as granular as they should be 🤔 
2. What's my responsibility when I'm assigned as a reviewer, but someone else has jumped in and started reviewing? Is my review still required or optional? (I tend to lean towards optional).

 I think in my ideal world automation would look something like:

- At least 1 person is assigned. Assignees are on the hook for reviewing the PR, or finding another person / removing themselves to reroll a reviewer if they feel unqualified to review the change. Additional assignees can be added by anyone, w/ the expectation that all assignees should lgtm/approve. At least 1 of these assignees should be a project OWNER.
- Reviewers aren't auto assigned by default, and are there mainly as a CC/FYI with no expectation of a sign off. 

(if anyone is familiar with Gerrit - tl;dr GitHub Assignee == Gerrit Reviewer, GitHub Reviewer == Gerrit CC)

**Reply by bobcatfish** [reply]
> and/or we put a manual burden on a small group of people to triage and assign PRs as they come in

my understanding is that this is much more what is already happening - or at least i believe @sbwsg does a lot of triaging (maybe build captains too?)

id love to have some automation around this - if it actually works and ppl are actually using it

**Comment by bobcatfish** [comment]
maybe backing this up with some stats, looking at pipelines [here is the 7 day moving average of reviews in the last 2 months](https://tekton.devstats.cd.foundation/d/46/pr-reviews-by-contributor?orgId=1&from=1609084415374&to=1614793628701&var-period=d7&var-repo_name=tektoncd%2Fpipeline&var-reviewers=%22dibyom%22&var-reviewers=%22bobcatfish%22&var-reviewers=%22PuneetPunamiya%22&var-reviewers=%22vinamra28%22&var-reviewers=%22sbwsg%22&var-reviewers=%22pritidesai%22&var-reviewers=%22wlynch%22&var-reviewers=%22vdemeester%22&var-reviewers=%22afrittoli%22&var-reviewers=%22chmouel%22&var-reviewers=%22savitaashture%22&var-reviewers=%22jerop%22&var-reviewers=%22piyush-garg%22&var-reviewers=%22nikhil-thomas%22&var-reviewers=%22pradeepitm12%22&var-reviewers=%22ImJasonH%22&var-reviewers=%22XinruZhang%22&var-reviewers=%22MarcelMue%22&var-reviewers=%22pratap0007%22&var-reviewers=%22SM43%22&var-reviewers=%22AlanGreene%22&var-reviewers=%22itewk%22&var-reviewers=%22Divyansh42%22&var-reviewers=%22rudeigerc%22&var-reviewers=%22skaegi%22&var-reviewers=%22gabemontero%22&var-reviewers=%22ziheng%22&var-reviewers=%22sergetron%22&var-reviewers=%22barthy1%22&var-reviewers=%22haidouks%22&var-reviewers=%22dprotaso%22&var-reviewers=%22GregDritschler%22&var-reviewers=%22chandanikumari%22&var-reviewers=%22souleb%22&var-reviewers=%22bigkevmcd%22&var-reviewers=%22dlorenc%22&var-reviewers=%22jmcshane%22&var-reviewers=%22yaoxiaoqi%22&var-reviewers=%22jlpettersson%22&var-reviewers=%22popcor255%22&var-reviewers=%22theofpa%22&var-reviewers=%22khrm%22&var-reviewers=%22NikeNano%22&var-reviewers=%22fiunchinho%22&var-reviewers=%22steveodonovan%22&var-reviewers=%22vincent-pli%22&var-reviewers=%22jromero%22&var-reviewers=%22zhangtbj%22&var-reviewers=%22spikeburton%22&var-reviewers=%22mogsie%22&var-reviewers=%22bahetiamit%22&var-reviewers=%22mattmoor%22&var-reviewers=%22a-rothwell%22&var-reviewers=%22karthikjeeyar%22&var-reviewers=%22rupalibehera%22&var-reviewers=%22FogDong%22&var-reviewers=%22cqbqdd11519%22&var-reviewers=%22ppitonak%22&var-reviewers=%22waveywaves%22&var-reviewers=%22pierretasci%22&var-reviewers=%22sthaha%22&var-reviewers=%22t-persson%22&var-reviewers=%22sbose78%22&var-reviewers=%22IamGabrielWu%22&var-reviewers=%22psschwei%22&var-reviewers=%22Hokwang%22&var-reviewers=%22riceluxs1t%22&var-reviewers=%2206kellyjac%22&var-reviewers=%22HeavyWombat%22&var-reviewers=%22djfinnoy%22&var-reviewers=%22hrishin%22&var-reviewers=%22kimsterv%22&var-reviewers=%22onthedock%22&var-reviewers=%22andsens%22&var-reviewers=%22httpsOmkar%22&var-reviewers=%22navidshaikh%22&var-reviewers=%22jjasghar%22&var-reviewers=%22iancoffey%22&var-reviewers=%22Huweicai%22&var-reviewers=%22ghatwala%22&var-reviewers=%2296RadhikaJadhav%22&var-reviewers=%22none%22)

![image](https://user-images.githubusercontent.com/432502/110020821-03a7e300-7cf8-11eb-8acd-baf5d65fdbd9.png)

maybe some other top reviewers can chime in on whether they are doing these reviews via what they are being assigned or something else? @pritidesai @jerop @afrittoli @sbwsg @GregDritschler @chmouel

**Reply by vdemeester** [reply]
Same on [6 month](https://tekton.devstats.cd.foundation/d/46/pr-reviews-by-contributor?orgId=1&from=now-6M&to=now&var-period=d7&var-repo_name=tektoncd%2Fpipeline&var-reviewers=%22dibyom%22&var-reviewers=%22bobcatfish%22&var-reviewers=%22PuneetPunamiya%22&var-reviewers=%22vinamra28%22&var-reviewers=%22sbwsg%22&var-reviewers=%22pritidesai%22&var-reviewers=%22wlynch%22&var-reviewers=%22vdemeester%22&var-reviewers=%22afrittoli%22&var-reviewers=%22chmouel%22&var-reviewers=%22savitaashture%22&var-reviewers=%22jerop%22&var-reviewers=%22piyush-garg%22&var-reviewers=%22nikhil-thomas%22&var-reviewers=%22pradeepitm12%22&var-reviewers=%22ImJasonH%22&var-reviewers=%22XinruZhang%22&var-reviewers=%22MarcelMue%22&var-reviewers=%22pratap0007%22&var-reviewers=%22SM43%22&var-reviewers=%22AlanGreene%22&var-reviewers=%22itewk%22&var-reviewers=%22Divyansh42%22&var-reviewers=%22rudeigerc%22&var-reviewers=%22skaegi%22&var-reviewers=%22gabemontero%22&var-reviewers=%22ziheng%22&var-reviewers=%22sergetron%22&var-reviewers=%22barthy1%22&var-reviewers=%22haidouks%22&var-reviewers=%22dprotaso%22&var-reviewers=%22GregDritschler%22&var-reviewers=%22chandanikumari%22&var-reviewers=%22souleb%22&var-reviewers=%22bigkevmcd%22&var-reviewers=%22dlorenc%22&var-reviewers=%22jmcshane%22&var-reviewers=%22yaoxiaoqi%22&var-reviewers=%22jlpettersson%22&var-reviewers=%22popcor255%22&var-reviewers=%22theofpa%22&var-reviewers=%22khrm%22&var-reviewers=%22NikeNano%22&var-reviewers=%22fiunchinho%22&var-reviewers=%22steveodonovan%22&var-reviewers=%22vincent-pli%22&var-reviewers=%22jromero%22&var-reviewers=%22zhangtbj%22&var-reviewers=%22spikeburton%22&var-reviewers=%22mogsie%22&var-reviewers=%22bahetiamit%22&var-reviewers=%22mattmoor%22&var-reviewers=%22a-rothwell%22&var-reviewers=%22karthikjeeyar%22&var-reviewers=%22rupalibehera%22&var-reviewers=%22FogDong%22&var-reviewers=%22cqbqdd11519%22&var-reviewers=%22ppitonak%22&var-reviewers=%22waveywaves%22&var-reviewers=%22pierretasci%22&var-reviewers=%22sthaha%22&var-reviewers=%22t-persson%22&var-reviewers=%22sbose78%22&var-reviewers=%22IamGabrielWu%22&var-reviewers=%22psschwei%22&var-reviewers=%22Hokwang%22&var-reviewers=%22riceluxs1t%22&var-reviewers=%2206kellyjac%22&var-reviewers=%22HeavyWombat%22&var-reviewers=%22djfinnoy%22&var-reviewers=%22hrishin%22&var-reviewers=%22kimsterv%22&var-reviewers=%22onthedock%22&var-reviewers=%22andsens%22&var-reviewers=%22httpsOmkar%22&var-reviewers=%22navidshaikh%22&var-reviewers=%22jjasghar%22&var-reviewers=%22iancoffey%22&var-reviewers=%22Huweicai%22&var-reviewers=%22ghatwala%22&var-reviewers=%2296RadhikaJadhav%22&var-reviewers=%22none%22) might be a bit better on usual reviewers.

I am not fully sure how to read those charts though. Is this taking into account reviews only (aka not single comment on a line or on the PR itself) ?

**Comment by ghost** [comment]
I look at every PR in Pipelines at least once just to check if I should review it in more detail or not.  I base that decision on the content of the PR and totally ignore the assigned Reviewers.

**Reply by jerop** [reply]
I do this too

**Comment by pritidesai** [comment]
The email notifications I receive (with requested review) at least draws my attention even in an area that I might not be most familiar with. After spending about 10 minutes on it and not fully understanding the changes (sometimes I end up asking dumb questions 😜 ), I push it to a back burner in my personal TODO and visit back if its open for more than a few days.

Apart from requested review, like @sbwsg and @jerop mentioned, every morning/afternoon I go through the pipeline repo for any new PRs and decide whether I should be reviewing it or not. I do ignore the assigned reviewers for most of the PRs but the `/cc` feature confuses me. Most of the times, I do not understand what that `/cc` mean? Is it that the folks cced are required to review before merging since they might have more insight into the changes? Or its just to draw their attention and its ok for non-cced folks to leave review and merge eventually? 

Thanks @bobcatfish for initiating this discussion?
